base/tps/src/main/java/org/dogtagpki/server/tps/rest/TPSCertService.java

```diff
@@ -162,6 +161,56 @@ public Response findCerts(String filter, String tokenID, Integer start, Integer
 
             return createOKResponse(response);
 
+        } catch (Exception e) {
+            throw new PKIException(e.getMessage());
+        }
+    }
+
+    private Response findAllCerts(String filter, Integer start, Integer size) {
+        String method = "TPSCertService:findAllCerts: ";
+        String msg = "";
+        TPSCertCollection response = new TPSCertCollection();
+
+        logger.debug("TPSCertService.findAllCerts({}, {}, {})", filter, start, size);
+
+        TPSEngine engine = TPSEngine.getInstance();
+        try {
+            TPSSubsystem subsystem = (TPSSubsystem) engine.getSubsystem(TPSSubsystem.ID);
+            TPSCertDatabase certDatabase = subsystem.getCertDatabase();
+            Iterator<TPSCertRecord> certRecs = certDatabase.findRecords(filter).iterator();
+            TokenDatabase tokenDatabase = subsystem.getTokenDatabase();
+            List<String> authorizedProfiles = getAuthorizedProfiles();
+            if (authorizedProfiles == null) {
+                msg = "authorizedProfiles null";
+                logger.debug("{}{}", method, msg);
+                throw new PKIException(method + msg);
+            }
+
+            int i = 0;
+            // skip to the start of the page
+            for (; i < start && certRecs.hasNext(); i++) certRecs.next();
+
+            // return entries up to the page size
+            for (; i < start + size && certRecs.hasNext(); i++) {
+                TPSCertRecord certRecord = certRecs.next();
+                String tokenID = certRecord.getTokenID();
+                TokenRecord tokenRecord = null;
+                try {
+                    tokenRecord = tokenDatabase.getRecord(tokenID);
+                } catch (Exception e) {
+                    // Proceed to next token if this one doesn't exist.
+                }
+                if (tokenRecord == null) continue;
+                String type = tokenRecord.getType();
+                if (!authorizedProfiles.contains(UserResource.ALL_PROFILES) && !authorizedProfiles.contains(type)) continue;
```

Comment:
Could `type` be `null`? In line 140 there is check.

Reply:
Yeah I forgot to reinstate it, I'll put it back before merging.